modules/commons/src/main/java/org/apache/ignite/schema/builder/Builder.java

```diff
@@ -0,0 +1,28 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.ignite.schema.builder;
+
+/**
+ * Builder base interface.
+ */
+public interface Builder {
+    /**
+     * @return Built object.
+     */
+    Object build();
```

Comment:
What if we parametrize this class, to avoid specifying 
"@" Override build(); 
method and replace it with 
extends Builder <HashIndex> {

Reply:
This will lead us to the hell of generics. 
Assume,  you have 2 classes A and B and 2 interfaces BuilderB extends BuilderA.
A try to generify builders via extending Builder<T> will make it impossible to reuse code in implementations BuilderB extends BuilderA even if B extends A.
Just try and you quickly end up with generics hell.
Templates like <? extends T> make things even worse.